Intermittent test failure - org.mozilla.fenix.home.recentbookmarks.RecentBookmarksFeatureTest - GIVEN no recent bookmarks WHEN feature starts THEN fetch bookmarks and notify store

### Firebase Test Run:
https://treeherder.mozilla.org/logviewer?job_id=344824917&repo=fenix&lineNumber=25940
### Stacktrace:
```
[task 2021-07-09T23:24:30.159Z]   TEST: GIVEN no recent bookmarks WHEN feature starts THEN fetch bookmarks and notify store
[task 2021-07-09T23:24:30.259Z]     I/glean/Dispatchers: Task queued for execution and delayed until flushed
[task 2021-07-09T23:24:30.259Z]     I/glean/Dispatchers: Task queued for execution and delayed until flushed
[task 2021-07-09T23:24:30.259Z]   FAILURE
[task 2021-07-09T23:24:30.360Z] 
[task 2021-07-09T23:24:30.360Z] java.lang.AssertionError: No action of type class org.mozilla.fenix.home.HomeFragmentAction$RecentBookmarksChange found
[task 2021-07-09T23:24:30.360Z] 	at mozilla.components.support.test.middleware.CaptureActionsMiddleware.findLastAction(CaptureActionsMiddleware.kt:44)
[task 2021-07-09T23:24:30.360Z] 	at mozilla.components.support.test.middleware.CaptureActionsMiddleware.assertLastAction(CaptureActionsMiddleware.kt:64)
[task 2021-07-09T23:24:30.360Z] 	at org.mozilla.fenix.home.recentbookmarks.RecentBookmarksFeatureTest$GIVEN no recent bookmarks WHEN feature starts THEN fetch bookmarks and notify store$1.invokeSuspend(RecentBookmarksFeatureTest.kt:85)
[task 2021-07-09T23:24:30.360Z] 	at org.mozilla.fenix.home.recentbookmarks.RecentBookmarksFeatureTest$GIVEN no recent bookmarks WHEN feature starts THEN fetch bookmarks and notify store$1.invoke(RecentBookmarksFeatureTest.kt)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.test.TestBuildersKt$runBlockingTest$deferred$1.invokeSuspend(TestBuilders.kt:50)
[task 2021-07-09T23:24:30.360Z] 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.test.TestCoroutineDispatcher.dispatch(TestCoroutineDispatcher.kt:50)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:305)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:27)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:109)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:158)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async(Builders.common.kt:91)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.BuildersKt.async(Unknown Source)
[task 2021-07-09T23:24:30.360Z] 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async$default(Builders.common.kt:84)
[task 2021-07-09T23:24:30.361Z] 	at kotlinx.coroutines.BuildersKt.async$default(Unknown Source)
[task 2021-07-09T23:24:30.361Z] 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest(TestBuilders.kt:49)
[task 2021-07-09T23:24:30.361Z] 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest(TestBuilders.kt:80)
[task 2021-07-09T23:24:30.361Z] 	at org.mozilla.fenix.home.recentbookmarks.RecentBookmarksFeatureTest.GIVEN no recent bookmarks WHEN feature starts THEN fetch bookmarks and notify store(RecentBookmarksFeatureTest.kt:67)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-07-09T23:24:30.361Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[task 2021-07-09T23:24:30.361Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.GeneratedMethodAccessor102.invoke(Unknown Source)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-07-09T23:24:30.361Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
[task 2021-07-09T23:24:30.361Z] 	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.GeneratedMethodAccessor101.invoke(Unknown Source)
[task 2021-07-09T23:24:30.361Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-07-09T23:24:30.361Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
[task 2021-07-09T23:24:30.361Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[task 2021-07-09T23:24:30.361Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[task 2021-07-09T23:24:30.361Z] 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
[task 2021-07-09T23:24:30.361Z] 	at java.lang.Thread.run(Thread.java:748)
```
### Build: 
https://github.com/mozilla-mobile/fenix/commit/76f24a7e132fb3d64e33e7606e9e6104cc04be5e

--- a/app/src/main/java/org/mozilla/fenix/home/recentbookmarks/RecentBookmarksFeature.kt
+++ b/app/src/main/java/org/mozilla/fenix/home/recentbookmarks/RecentBookmarksFeature.kt
@@ -4,6 +4,7 @@
 
 package org.mozilla.fenix.home.recentbookmarks
 
+import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.Job
@@ -22,16 +23,18 @@ import org.mozilla.fenix.home.HomeFragmentStore
  *  @param bookmarksUseCase the [BookmarksUseCase] for retrieving the list of recently saved
 *   bookmarks from storage.
  *  @param scope the [CoroutineScope] used to fetch the bookmarks list
+ *  @param ioDispatcher the [CoroutineDispatcher] for performing read/write operations.
  */
 class RecentBookmarksFeature(
     private val homeStore: HomeFragmentStore,
     private val bookmarksUseCase: BookmarksUseCase,
-    private val scope: CoroutineScope
+    private val scope: CoroutineScope,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
 ) : LifecycleAwareFeature {
     internal var job: Job? = null
 
     override fun start() {
-        job = scope.launch(Dispatchers.IO) {
+        job = scope.launch(ioDispatcher) {
             val bookmarks = bookmarksUseCase.retrieveRecentBookmarks()
 
             homeStore.dispatch(HomeFragmentAction.RecentBookmarksChange(bookmarks))

--- a/app/src/test/java/org/mozilla/fenix/home/recentbookmarks/RecentBookmarksFeatureTest.kt
+++ b/app/src/test/java/org/mozilla/fenix/home/recentbookmarks/RecentBookmarksFeatureTest.kt
@@ -7,11 +7,9 @@ package org.mozilla.fenix.home.recentbookmarks
 import io.mockk.coEvery
 import io.mockk.coVerify
 import io.mockk.mockk
-import io.mockk.spyk
-import io.mockk.verify
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.test.TestCoroutineDispatcher
-import kotlinx.coroutines.test.TestCoroutineScope
 import kotlinx.coroutines.test.runBlockingTest
 import mozilla.components.concept.storage.BookmarkNode
 import mozilla.components.concept.storage.BookmarkNodeType
@@ -20,8 +18,6 @@ import mozilla.components.support.test.middleware.CaptureActionsMiddleware
 import mozilla.components.support.test.rule.MainCoroutineRule
 import org.junit.After
 import org.junit.Assert.assertEquals
-import org.junit.Assert.assertNotNull
-import org.junit.Assert.assertNull
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
@@ -36,7 +32,6 @@ class RecentBookmarksFeatureTest {
     private val middleware = CaptureActionsMiddleware<HomeFragmentState, HomeFragmentAction>()
     private val homeStore = HomeFragmentStore(middlewares = listOf(middleware))
     private val bookmarksUseCases: BookmarksUseCase = mockk(relaxed = true)
-    private val scope = TestCoroutineScope()
     private val testDispatcher = TestCoroutineDispatcher()
     private val bookmark = BookmarkNode(
         type = BookmarkNodeType.ITEM,
@@ -58,7 +53,6 @@ class RecentBookmarksFeatureTest {
 
     @After
     fun cleanUp() {
-        scope.cleanupTestCoroutines()
         testDispatcher.cleanupTestCoroutines()
     }
 
@@ -68,12 +62,13 @@ class RecentBookmarksFeatureTest {
             val feature = RecentBookmarksFeature(
                 homeStore,
                 bookmarksUseCases,
-                scope
+                CoroutineScope(testDispatcher),
+                testDispatcher
             )
 
-            feature.start()
-
             assertEquals(emptyList<BookmarkNode>(), homeStore.state.recentBookmarks)
+
+            feature.start()
 
             testDispatcher.advanceUntilIdle()
             homeStore.waitUntilIdle()
@@ -86,23 +81,4 @@ class RecentBookmarksFeatureTest {
                 assertEquals(listOf(bookmark), it.recentBookmarks)
             }
         }
-
-    @Test
-    fun `WHEN the feature is destroyed THEN the job is cancelled`() {
-        val feature = spyk(RecentBookmarksFeature(
-            homeStore,
-            bookmarksUseCases,
-            scope
-        ))
-
-        assertNull(feature.job)
-
-        feature.start()
-
-        assertNotNull(feature.job)
-
-        feature.stop()
-
-        verify(exactly = 1) { feature.stop() }
-    }
 }

